OpenRA.Game/Network/Order.cs

```diff
@@ -247,91 +249,108 @@ public Order(string orderString, Actor subject, Target target, bool queued)
-			}
+					if (TargetString != null)
+						fields |= OrderFields.TargetString;
 
-			w.Write((byte)0xFF);
-			w.Write(OrderString);
-			w.Write(UIntFromActor(Subject));
+					if (ExtraData != 0)
+						fields |= OrderFields.ExtraData;
 
-			if (Target.SerializableType != TargetType.Invalid)
-				fields |= OrderFields.Target;
+					if (Target.SerializableType != TargetType.Invalid)
+						fields |= OrderFields.Target;
 
-			if (Queued)
-				fields |= OrderFields.Queued;
+					if (Queued)
+						fields |= OrderFields.Queued;
 
-			if (ExtraLocation != CPos.Zero)
-				fields |= OrderFields.ExtraLocation;
+					if (ExtraLocation != CPos.Zero)
+						fields |= OrderFields.ExtraLocation;
 
-			if (Target.SerializableCell != null)
-				fields |= OrderFields.TargetIsCell;
+					if (Target.SerializableCell != null)
+						fields |= OrderFields.TargetIsCell;
 
-			w.Write((byte)fields);
+					w.Write((byte)fields);
 
-			if (fields.HasField(OrderFields.Target))
-			{
-				w.Write((byte)Target.SerializableType);
-				switch (Target.SerializableType)
-				{
-					case TargetType.Actor:
-						w.Write(UIntFromActor(Target.SerializableActor));
-						break;
-					case TargetType.FrozenActor:
-						w.Write(Target.FrozenActor.Viewer.PlayerActor.ActorID);
-						w.Write(Target.FrozenActor.ID);
-						break;
-					case TargetType.Terrain:
-						if (fields.HasField(OrderFields.TargetIsCell))
+					if (fields.HasField(OrderFields.Subject))
+						w.Write(UIntFromActor(Subject));
+
+					if (fields.HasField(OrderFields.Target))
+					{
+						w.Write((byte)Target.SerializableType);
+						switch (Target.SerializableType)
 						{
-							w.Write(Target.SerializableCell.Value);
-							w.Write((byte)Target.SerializableSubCell);
+							case TargetType.Actor:
+								w.Write(UIntFromActor(Target.SerializableActor));
+								break;
+							case TargetType.FrozenActor:
+								w.Write(Target.FrozenActor.Viewer.PlayerActor.ActorID);
+								w.Write(Target.FrozenActor.ID);
+								break;
+							case TargetType.Terrain:
+								if (fields.HasField(OrderFields.TargetIsCell))
+								{
+									w.Write(Target.SerializableCell.Value);
+									w.Write((byte)Target.SerializableSubCell);
+								}
+								else
+									w.Write(Target.SerializablePos);
+								break;
 						}
-						else
-							w.Write(Target.SerializablePos);
-						break;
-				}
-			}
+					}
 
-			if (fields.HasField(OrderFields.TargetString))
-				w.Write(TargetString);
+					if (fields.HasField(OrderFields.TargetString))
+						w.Write(TargetString);
 
-			if (fields.HasField(OrderFields.ExtraLocation))
-				w.Write(ExtraLocation);
+					if (fields.HasField(OrderFields.ExtraLocation))
+						w.Write(ExtraLocation);
 
-			if (fields.HasField(OrderFields.ExtraData))
-				w.Write(ExtraData);
+					if (fields.HasField(OrderFields.ExtraData))
+						w.Write(ExtraData);
+
+					break;
+				}
+
+				default:
+					throw new NotImplementedException("Cannot serialize order type {0}".F(Type));
```

Comment:
Small nitpick: the intention of throwing could be "lost" by using NotImplementedException